src/c_ext/hamt/node.hpp

```diff
@@ -0,0 +1,518 @@
+            this->size = 0;
+            this->val = T();
+            this->make_mut(false);
+
+            this->chdrn = new vector<node<T> * >();
+        }
+
+        node(const node<T>& rhs)
+        {
+            if (this != &rhs) {
+                this->type = rhs.get_type();
+                this->hash = rhs.get_hash();
+                this->key = rhs.get_key();
+                this->mask = rhs.get_mask();
+                this->size = rhs.get_size();
+                this->val = rhs.get_val();
+                this->make_mut(rhs.editable());
+
+                this->chdrn = rhs.get_chdrn();
+            }
+        }
+
+        node<T> &operator=(const node<T> &rhs)
+        {
+            if (this == &rhs) {
+                return *this;
+            }
+
+            this->type = rhs.get_type();
+            this->hash = rhs.get_hash();
+            this->key = rhs.get_key();
+            this->mask = rhs.get_mask();
+            this->size = rhs.get_size();
+            this->val = rhs.get_val();
+            this->make_mut(rhs.editable());
+
+            this->chdrn = rhs.get_chdrn();
+
+            return *this;
+        }
+
+        bool operator==(const node<T> &rhs) const {
+            if (this == &rhs)
+                return true;
+
+            return (this->type == rhs.get_type()) &&
+                    (this->hash == rhs.get_hash()) &&
+                    (this->key == rhs.get_key()) &&
+                    (this->mask == rhs.get_mask()) &&
+                    (this->size == rhs.get_size()) &&
+                    (this->val == rhs.get_val()) &&
+                    (this->editable() == rhs.editable()) &&
+                    (this->chdrn == rhs.get_chdrn()) &&
+                    (*this->chdrn == *rhs.get_chdrn());
+        }
+
+        bool operator!=(const node<T> &rhs) const {
+            return !(*this == rhs);
+        }
+/*
+        node<T>(node<T>&& rhs) {
+            this->type = rhs.type;
+            this->hash = rhs.hash;
+            this->key = rhs.key;
+            this->mask = rhs.mask;
+            this->size = 0;
+            this->val = rhs.val;
+
+            this->chdrn = std::move(rhs.chdrn);
+        }
+*/
+        ~node<T>() {}
+
+        types get_type() const { return this->type; }
+
+        u_int get_hash() const { return this->hash; }
+
+        string get_key() const { return this->key; }
+
+        u_int get_mask() const { return this->mask; }
+
+        void set_size(const int size) { this->size = size; }
+
+        int get_size() const { return this->size; }
+
+        T get_val() const { return this->val; }
+
+        bool editable() const { return this->can_edit; }
+
+        void make_mut(const bool mode) { this->can_edit = mode; }
+
+        void set_chdrn(vector<node<T> * > *chdrn) { this->chdrn = chdrn; }
+
+        vector<node<T> * > *get_chdrn() const { return this->chdrn; }
+
+        virtual node<T> *modify(const int shift, T(*fun)(T), const u_int hash,
```

Comment:
there was a compilation issue i had to resolve
this is overwritten by its subclasses